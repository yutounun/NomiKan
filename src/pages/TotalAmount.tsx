import React, { useState } from "react";
import {
  Alert,
  Box,
  Snackbar,
  Stack,
  Step,
  StepLabel,
  Stepper,
  Typography,
} from "@mui/material";

import BaseLayout from "components/Commons/Layout/BaseLayout";
import MyHeader from "components/Commons/Organisms/MyHeader";

import MyButton from "components/Commons/Atoms/MyButton";
import { useCostStore } from "stores/cost";
import CostTextField from "components/TotalAmount/Atoms/CostTextField";

function TotalAmount() {
  const [cost, setCost] = useCostStore((state) => [state.cost, state.setCost]);
  const [open, setOpen] = useState(false);
  const [localCost, setLocalCost] = useState<number | string | undefined>("");

  /** save cost to Zustand store */
  const handleClickButton = () => {
    // set input cost on Session storage
    setCost(localCost);

    // open the snack bar
    setOpen(true);

    // clear input cost
    setLocalCost("");
  };

  /** close the snack bar */
  const handleOnClose = () => {
    setOpen(false);
  };

  /** handle input cost */
  const handleUpdateCost = (updatedCost: number | string | undefined) => {
    setLocalCost(updatedCost);
  };

  const steps = ["合計金額", "メンバー追加", "割合入力", "結果"];
  return (
    <BaseLayout>
      <Stack direction="column" justifyContent="center" alignItems="center">
        <MyHeader value="合計金額" />

        <Box sx={{ width: "100%" }} mt={10}>
          <Stepper activeStep={0} alternativeLabel>
            {steps.map((label) => (
              <Step key={label}>
                <StepLabel>
                  <Typography variant="body2">{label}</Typography>
                </StepLabel>
              </Step>
            ))}
          </Stepper>
        </Box>
        <Stack
          direction="column"
          justifyContent="center"
          mt={10}
          alignItems="center"
          gap="70px"
        >
          <Box>{localCost}</Box>
          <CostTextField onChangeCost={handleUpdateCost} value={localCost} />
          <MyButton
            onClick={handleClickButton}
            disabled={localCost === ""}
            value="登録"
            variant="contained"
            sx={{ borderRadius: "30px", width: "100%" }}
          />
        </Stack>
      </Stack>

      {/* Alert for the completion of registration  */}
      <Snackbar
        anchorOrigin={{ vertical: "bottom", horizontal: "center" }}
        open={open}
        onClose={handleOnClose}
      >
        <Alert onClose={handleOnClose} severity="info" sx={{ width: "100%" }}>
          登録しました！
        </Alert>
      </Snackbar>
    </BaseLayout>
  );
}

export default TotalAmount;
