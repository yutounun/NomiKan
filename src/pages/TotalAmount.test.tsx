import { fireEvent, render, screen, cleanup } from "@testing-library/react";
import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import TotalAmount from "./TotalAmount";

/**
 * Requirements
 * 1. The component must be rendered in the DOM
 * 2. Stepper component must be rendered and have all the steps
 * 3. Input component must be rendered with a placeholder
 * 4. Alert component must be rendered and have a message after the completion of registration
 * 5. Button component must be rendered
 * 6. Button must not be clicked when the input is empty
 * 7. Active step on Stepper must be 0
 * 8. Cost must be saved on Session storage, when the button is clicked
 */

describe("TotalAmount", () => {
  afterEach(() => cleanup());

  it("doesn't allow user to click Button when Input is empty", () => {
    render(
      <Router>
        <TotalAmount />
      </Router>
    );
    expect(screen.getByRole("button", { name: /Submit/i })).toBeDisabled();
  });

  it("renders correctly", () => {
    render(
      <Router>
        <TotalAmount />
      </Router>
    );
  });

  it("takes a snapshot", () => {
    const { asFragment } = render(
      <Router>
        <TotalAmount />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders all steps in the Stepper component", () => {
    render(
      <Router>
        <TotalAmount />
      </Router>
    );
    expect(screen.getByTestId("total amount")).toBeInTheDocument();
    expect(screen.getByTestId("members")).toBeInTheDocument();
    expect(screen.getByTestId("ratio")).toBeInTheDocument();
    expect(screen.getByTestId("result")).toBeInTheDocument();
  });

  it("renders an Input component with a placeholder", () => {
    render(
      <Router>
        <TotalAmount />
      </Router>
    );
    const inputElem = screen.getByLabelText("Enter the total cost");
    expect(inputElem).toBeInTheDocument();
  });

  it("does not render alert at first", () => {
    render(
      <Router>
        <TotalAmount />
      </Router>
    );
    expect(screen.queryByRole("alert")).toBeFalsy();
  });

  it("renders an Alert component after registration", () => {
    render(
      <Router>
        <TotalAmount />
      </Router>
    );

    // Type cost in the textbox field
    fireEvent.change(
      screen.getByRole("spinbutton", {
        name: /Enter the total cost/i,
      }),
      { target: { value: "500" } }
    );

    // Click registration button
    fireEvent(
      screen.getByRole("button", { name: /Submit/i }),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(screen.getByText("Successfully completed registration!!")).toBeInTheDocument();
  });

  it("renders a Button component", () => {
    render(
      <Router>
        <TotalAmount />
      </Router>
    );
    const buttonElem = screen.getByRole("spinbutton", {
      name: /Enter the total cost/i,
    });
    expect(buttonElem).toBeInTheDocument();
  });

  it("sets the Active step on Stepper to be 0", () => {
    render(
      <Router>
        <TotalAmount />
      </Router>
    );
    const stepperElem = screen.getByTestId("total amount");

    // Only inactive steps have "Mui-disabled" class
    expect(stepperElem).not.toHaveClass("Mui-disabled");
  });

  it("saves cost on session storage when Button is clicked", () => {
    render(
      <Router>
        <TotalAmount />
      </Router>
    );

    // Type cost in the textbox field
    fireEvent.change(
      screen.getByRole("spinbutton", {
        name: /Enter the total cost/i,
      }),
      { target: { value: "500" } }
    );

    // Click registration button
    fireEvent(
      screen.getByRole("button", { name: /Submit/i }),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(window.sessionStorage.getItem("nomikan-storage")).toEqual(
      "{\"state\":{\"cost\":\"500\",\"members\":[]},\"version\":0}"
    );
  });
});
