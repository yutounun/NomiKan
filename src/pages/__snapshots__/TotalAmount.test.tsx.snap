// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TotalAmount takes a snapshot 1`] = `
<DocumentFragment>
  <main
    data-testid="main"
  >
    <div
      class="MuiStack-root css-w004l2-MuiStack-root"
    >
      <div
        class="MuiStack-root css-wzdyko-MuiStack-root"
        data-testid="header"
      >
        <h1
          class="MuiTypography-root MuiTypography-h1 css-1doh40h-MuiTypography-root"
        >
          Total Amount
        </h1>
      </div>
      <div
        class="MuiStack-root css-1rh9169-MuiStack-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArrowBackIosIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.67 3.87 9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
          />
        </svg>
      </div>
      <div
        class="MuiBox-root css-jxhp8u"
      >
        <div
          class="MuiStepper-root MuiStepper-horizontal MuiStepper-alternativeLabel css-10mg1vw-MuiStepper-root"
          data-testid="stepper"
        >
          <div
            class="MuiStep-root MuiStep-horizontal MuiStep-alternativeLabel css-n7tliy-MuiStep-root"
          >
            <span
              class="MuiStepLabel-root MuiStepLabel-horizontal MuiStepLabel-alternativeLabel css-ascpo7-MuiStepLabel-root"
              data-testid="total amount"
            >
              <span
                class="MuiStepLabel-iconContainer Mui-active MuiStepLabel-alternativeLabel css-vnkopk-MuiStepLabel-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root Mui-active css-1u4zpwo-MuiSvgIcon-root-MuiStepIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="12"
                  />
                  <text
                    class="MuiStepIcon-text css-117w1su-MuiStepIcon-text"
                    dominant-baseline="central"
                    text-anchor="middle"
                    x="12"
                    y="12"
                  >
                    1
                  </text>
                </svg>
              </span>
              <span
                class="MuiStepLabel-labelContainer MuiStepLabel-alternativeLabel css-1vyamtt-MuiStepLabel-labelContainer"
              >
                <span
                  class="MuiStepLabel-label Mui-active MuiStepLabel-alternativeLabel css-1hv8oq8-MuiStepLabel-label"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                  >
                    total amount
                  </p>
                </span>
              </span>
            </span>
          </div>
          <div
            class="MuiStep-root MuiStep-horizontal MuiStep-alternativeLabel css-n7tliy-MuiStep-root"
          >
            <div
              class="MuiStepConnector-root MuiStepConnector-horizontal MuiStepConnector-alternativeLabel Mui-disabled css-zpcwqm-MuiStepConnector-root"
            >
              <span
                class="MuiStepConnector-line MuiStepConnector-lineHorizontal css-z7uhs0-MuiStepConnector-line"
              />
            </div>
            <span
              class="MuiStepLabel-root MuiStepLabel-horizontal Mui-disabled MuiStepLabel-alternativeLabel css-ascpo7-MuiStepLabel-root"
              data-testid="members"
            >
              <span
                class="MuiStepLabel-iconContainer Mui-disabled MuiStepLabel-alternativeLabel css-vnkopk-MuiStepLabel-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root css-1u4zpwo-MuiSvgIcon-root-MuiStepIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="12"
                  />
                  <text
                    class="MuiStepIcon-text css-117w1su-MuiStepIcon-text"
                    dominant-baseline="central"
                    text-anchor="middle"
                    x="12"
                    y="12"
                  >
                    2
                  </text>
                </svg>
              </span>
              <span
                class="MuiStepLabel-labelContainer MuiStepLabel-alternativeLabel css-1vyamtt-MuiStepLabel-labelContainer"
              >
                <span
                  class="MuiStepLabel-label Mui-disabled MuiStepLabel-alternativeLabel css-1hv8oq8-MuiStepLabel-label"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                  >
                    members
                  </p>
                </span>
              </span>
            </span>
          </div>
          <div
            class="MuiStep-root MuiStep-horizontal MuiStep-alternativeLabel css-n7tliy-MuiStep-root"
          >
            <div
              class="MuiStepConnector-root MuiStepConnector-horizontal MuiStepConnector-alternativeLabel Mui-disabled css-zpcwqm-MuiStepConnector-root"
            >
              <span
                class="MuiStepConnector-line MuiStepConnector-lineHorizontal css-z7uhs0-MuiStepConnector-line"
              />
            </div>
            <span
              class="MuiStepLabel-root MuiStepLabel-horizontal Mui-disabled MuiStepLabel-alternativeLabel css-ascpo7-MuiStepLabel-root"
              data-testid="ratio"
            >
              <span
                class="MuiStepLabel-iconContainer Mui-disabled MuiStepLabel-alternativeLabel css-vnkopk-MuiStepLabel-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root css-1u4zpwo-MuiSvgIcon-root-MuiStepIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="12"
                  />
                  <text
                    class="MuiStepIcon-text css-117w1su-MuiStepIcon-text"
                    dominant-baseline="central"
                    text-anchor="middle"
                    x="12"
                    y="12"
                  >
                    3
                  </text>
                </svg>
              </span>
              <span
                class="MuiStepLabel-labelContainer MuiStepLabel-alternativeLabel css-1vyamtt-MuiStepLabel-labelContainer"
              >
                <span
                  class="MuiStepLabel-label Mui-disabled MuiStepLabel-alternativeLabel css-1hv8oq8-MuiStepLabel-label"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                  >
                    ratio
                  </p>
                </span>
              </span>
            </span>
          </div>
          <div
            class="MuiStep-root MuiStep-horizontal MuiStep-alternativeLabel css-n7tliy-MuiStep-root"
          >
            <div
              class="MuiStepConnector-root MuiStepConnector-horizontal MuiStepConnector-alternativeLabel Mui-disabled css-zpcwqm-MuiStepConnector-root"
            >
              <span
                class="MuiStepConnector-line MuiStepConnector-lineHorizontal css-z7uhs0-MuiStepConnector-line"
              />
            </div>
            <span
              class="MuiStepLabel-root MuiStepLabel-horizontal Mui-disabled MuiStepLabel-alternativeLabel css-ascpo7-MuiStepLabel-root"
              data-testid="result"
            >
              <span
                class="MuiStepLabel-iconContainer Mui-disabled MuiStepLabel-alternativeLabel css-vnkopk-MuiStepLabel-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root css-1u4zpwo-MuiSvgIcon-root-MuiStepIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="12"
                  />
                  <text
                    class="MuiStepIcon-text css-117w1su-MuiStepIcon-text"
                    dominant-baseline="central"
                    text-anchor="middle"
                    x="12"
                    y="12"
                  >
                    4
                  </text>
                </svg>
              </span>
              <span
                class="MuiStepLabel-labelContainer MuiStepLabel-alternativeLabel css-1vyamtt-MuiStepLabel-labelContainer"
              >
                <span
                  class="MuiStepLabel-label Mui-disabled MuiStepLabel-alternativeLabel css-1hv8oq8-MuiStepLabel-label"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                  >
                    result
                  </p>
                </span>
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        class="MuiStack-root css-18tvtp4-MuiStack-root"
      >
        <div
          class="MuiStack-root css-1q3ulhk-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1gvvk8k-MuiTypography-root"
          >
            Â¥
          </p>
          <div
            class="MuiFormControl-root MuiTextField-root css-4qha4g-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r2:"
              id=":r2:-label"
            >
              Enter the total cost
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r2:"
                inputmode="numeric"
                pattern="\\\\d*"
                type="number"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Enter the total cost
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1hby8u7-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="DoneOutlineIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m19.77 5.03 1.4 1.4L8.43 19.17l-5.6-5.6 1.4-1.4 4.2 4.2L19.77 5.03m0-2.83L8.43 13.54l-4.2-4.2L0 13.57 8.43 22 24 6.43 19.77 2.2z"
              />
            </svg>
          </span>
          Submit
        </button>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
